File: src/main/java/core/procedures/io/ReadChar.java
package core.procedures.io;

import core.Repl;
import core.exceptions.SCMIOException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.InputPort;

import java.io.IOException;

public final class ReadChar extends AFn {

  public ReadChar() {
    super(new FnArgsBuilder().max(1).rest(InputPort.class).build());
  }

  @Override
  public String getName() {
    return "read-char";
  }

  @Override
  public Object apply(Object... args) {
    InputPort inputPort;
    if (args.length == 0) {
      inputPort = Repl.getCurrentInputPort();
    } else {
      inputPort = ((InputPort)args[0]);
    }
    try {
      return (char)inputPort.read();
    } catch (IOException e) {
      throw new SCMIOException(e);
    }
  }
}

File: src/main/java/core/procedures/io/ReadChar.kt
package core.procedures.io

import core.Repl
import core.procedures.AFn
import core.procedures.Arity.Range
import core.scm.InputPort

class ReadChar : AFn<Any?, Char>(name = "read-char", arity = Range(0, 1), restArgsType = InputPort::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        0    -> Repl.currentInputPort
        else -> args[0] as InputPort
    }.read().toChar()
}
